### PR TITLE
Add CKAN support

### PR DESCRIPTION
**Pull request by Joonalai**
This proposal adds support for CKAN format in addition to used CSW format.

Lot of refactoring has been made and therefore the major version number is incremented to 1.0-SNAPSHOT.

### CODE REVIEW

File: pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/GeographicBoundingBoxCustomExtractor.java [hunk 1]
```diff
@@ -26,9 +28,14 @@ private Double getDoubleValue(String elementName, XPath xPath, Node node)
 	}
 	
 	@Override
-	public Object process(XPath xPath, Node node) throws XPathException {
+	public Object process(DocumentQuerier documentQuerier, QueryResult queryResult) throws XPathException {
 		Double [] ret = new Double[4];
-		
+		if (!(documentQuerier instanceof XmlDocumentQuerierImpl)) {
```
[the remark this answers is not in the document]
Comment (sampov2) [reply pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/GeographicBoundingBoxCustomExtractor.java]: Or throw a more descriptive exception as discussed in the ResponsiblePartyExtractor

File: pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/FieldExtractorConfigurationImpl.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		return customNodeListExtractor;
+	}
+
+	public void setCustomNodeListExtractor(ListCustomExtractor customNodeListExtractor) {
+		this.customNodeListExtractor = customNodeListExtractor;
+	}
+
+	@Override
+	public Object process(Document doc, DocumentQuerier documentQuerier) throws DocumentProcessingException {
+		List<QueryResult> nodeList = documentQuerier.process(getQuery(), doc);
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/FieldExtractorConfigurationImpl.java]: Done!

File: pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/JsonDocument.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package fi.maanmittauslaitos.pta.search.documentprocessor;
+
+public class JsonDocument extends Document {
+	private String content;
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/JsonDocument.java]: Good idea. Done.

File: pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/query/DocumentQuerier.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package fi.maanmittauslaitos.pta.search.documentprocessor.query;
+
+
+import fi.maanmittauslaitos.pta.search.documentprocessor.Document;
+import fi.maanmittauslaitos.pta.search.documentprocessor.DocumentProcessingException;
+
+import java.util.List;
+
+public interface DocumentQuerier {
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/query/DocumentQuerier.java]: Done.

File: pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/query/JsonDocumentQuerierImpl.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+	@Override
+	public List<QueryResult> process(String query, Document document) throws DocumentProcessingException {
+		if (!(document instanceof JsonDocument)) {
+			throw new DocumentProcessingException("Document was not instance of JsonDocument");
+		}
+		List<QueryResult> results;
+		try {
+			results = JsonPath.using(configuration)
+					.parse(((JsonDocument) document).getContent())
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/query/JsonDocumentQuerierImpl.java]: Done.

File: pta-intelligent-search-harvester/src/main/java/fi/maanmittauslaitos/pta/search/source/json/CKANHarvesterSource.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		private void getNextRow() {
+			int startPosition = numberOfRowsProcessed;
+			int maxRows = getBatchSize();
+			logger.debug("Requesting records startPosition = " + startPosition + ",maxRecords = " + maxRows);
+
+			StringBuilder reqUrl = new StringBuilder(getOnlineResource());
+			reqUrl.append("?");
+			reqUrl.append(String.format("q=%s", getQuery()));
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-harvester/src/main/java/fi/maanmittauslaitos/pta/search/source/json/CKANHarvesterSource.java]: Done.

File: pta-intelligent-search-harvester/src/main/java/fi/maanmittauslaitos/pta/search/source/json/CKANHarvesterSource.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+			reqUrl.append(String.format("&rows=%s&start=%s", maxRows, startPosition));
+
+			logger.trace("JSON GetRecords URL: " + reqUrl);
+
+			try (InputStream is = new URL(reqUrl.toString()).openStream()) {
+				JsonNode response = objectMapper.readTree(is);
+				if (response.has("success") && response.get("success").booleanValue()) {
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-harvester/src/main/java/fi/maanmittauslaitos/pta/search/source/json/CKANHarvesterSource.java]: Added extra `&& response.get("success") != null`.

File: pta-intelligent-search-metadata-extractor/pom.xml [hunk 8]
```diff
@@ -23,6 +23,16 @@
 			<artifactId>log4j-api</artifactId>
 			<version>2.7</version>
 		</dependency>
+		<dependency>
+			<groupId>com.jayway.jsonpath</groupId>
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-metadata-extractor/pom.xml]: Dependencies was not needed in the extractor after moving the configuration to processor module as well. However, Jackson dependency had to be added to processor module.

File: pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/GeographicBoundingBoxCustomExtractor.java [hunk 9]
```diff
@@ -1,19 +1,21 @@
...
 import javax.xml.xpath.XPathException;
 import javax.xml.xpath.XPathExpression;
+import java.util.Arrays;
 
-import org.w3c.dom.Node;
-
-import fi.maanmittauslaitos.pta.search.documentprocessor.XPathCustomExtractor;
-
-public class GeographicBoundingBoxXPathCustomExtractor implements XPathCustomExtractor {
+public class GeographicBoundingBoxCustomExtractor implements CustomExtractor {
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/GeographicBoundingBoxCustomExtractor.java]: Done.

File: pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/GeographicBoundingBoxCustomExtractor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/GeographicBoundingBoxCustomExtractor.java]: Done.

File: pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/ResponsiblePartyCustomExtractor.java [hunk 10]
```diff
@@ -15,8 +20,8 @@
 
 import static fi.maanmittauslaitos.pta.search.metadata.utils.XPathHelper.matches;
 
-public class ResponsiblePartyXPathCustomExtractor implements XPathCustomExtractor {
-	private static Logger logger = LogManager.getLogger(ResponsiblePartyXPathCustomExtractor.class);
+public class ResponsiblePartyCustomExtractor implements CustomExtractor {
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/ResponsiblePartyCustomExtractor.java]: Done.

File: pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/ResponsiblePartyCustomExtractor.java [hunk 11]
```diff
@@ -38,10 +43,17 @@ public void setOrganisationNameRewriter(TextRewriter organisationNameRewriter) {
...
 		return organisationNameRewriter;
 	}
-	
-	
+
 	@Override
-	public Object process(XPath xPath, Node node) throws XPathException {
+	public Object process(DocumentQuerier documentQuerier, QueryResult queryResult) throws XPathException, DocumentProcessingException {
+		if (!(documentQuerier instanceof XmlDocumentQuerierImpl)) {
+			throw new DocumentProcessingException("documentQuerier was not instance of XmlDocumentQuerierImpl");
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-metadata-extractor/src/main/java/fi/maanmittauslaitos/pta/search/metadata/ResponsiblePartyCustomExtractor.java]: I agree. I changed the clause text to be more descriptive.

File: pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/DocumentProcessorFactory.java [hunk 12]
```diff
@@ -1,101 +1,35 @@
...
-		
+
+		return instance;
+	}
+
+	public DocumentProcessor createXmlProcessor(DocumentProcessingConfiguration configuration) throws ParserConfigurationException {
+		return new XmlDocumentProcessorImpl(configuration);
+	}
+
+	public DocumentProcessor createJsonProcessor(DocumentProcessingConfiguration configuration, Configuration jsonPathConfiguration) throws ParserConfigurationException {
```
[the remark this answers is not in the document]
Comment (Joonalai) [reply pta-intelligent-search-document-processor/src/main/java/fi/maanmittauslaitos/pta/search/documentprocessor/DocumentProcessorFactory.java]: I was not sure where to initialize the JsonPath configuration, but I think that JsonDocumentProcessorImpl is better place for that and therefore the parameters could stay the same.

**Pull request merged**